flink-state-backends/flink-statebackend-heap-spillable/src/main/java/org/apache/flink/runtime/state/heap/CopyOnWriteSkipListStateMap.java

```diff
@@ -472,6 +397,74 @@ S putNode(ByteBuffer keyByteBuffer, int keyOffset, int keyLen, byte[] value, boo
 	 * @return the old state. Null will be returned if key does not exist or returnOldState is false.
 	 */
 	private S removeNode(ByteBuffer keyByteBuffer, int keyOffset, int keyLen, boolean returnOldState) {
+		Tuple4<Long, Long, Boolean, S> result = iterateAndProcess(keyByteBuffer, keyOffset, keyLen,
+			(tuple3, isRemoved) -> {
+				long prevNode = tuple3.f0;
+				long currentNode = tuple3.f1;
+				long nextNode = tuple3.f2;
+				// if the node has been logically removed, and can not be physically
+				// removed here, just return null
+				if (isRemoved && highestRequiredSnapshotVersion != 0) {
+					return null;
+				}
+
+				long oldValuePointer;
+				boolean oldValueNeedFree;
+
+				if (highestRequiredSnapshotVersion == 0) {
+					// do physically remove only when there is no snapshot running
+					oldValuePointer = doPhysicalRemoveAndGetValue(currentNode, prevNode, nextNode);
+					// the node has been logically removed, and remove it from the set
+					if (isRemoved) {
+						logicallyRemovedNodes.remove(currentNode);
+					}
+					oldValueNeedFree = true;
+				} else {
+					int version = SkipListUtils.helpGetNodeLatestVersion(currentNode, spaceAllocator);
+					if (version < highestRequiredSnapshotVersion) {
+						// the newest-version value may be used by snapshots, and update it with copy-on-write
+						oldValuePointer = updateValueWithCopyOnWrite(currentNode, null);
+						oldValueNeedFree = false;
+					} else {
+						// replace the newest-version value.
+						oldValuePointer = updateValueWithReplace(currentNode, null);
+						oldValueNeedFree = true;
+					}
+
+					helpSetNodeStatus(currentNode, NodeStatus.REMOVE);
+					logicallyRemovedNodes.add(currentNode);
+				}
+
+				S oldState = null;
+				if (returnOldState) {
+					oldState = helpGetState(oldValuePointer);
+				}
+
+				if (oldValueNeedFree) {
+					spaceAllocator.free(oldValuePointer);
+				}
+
+				return oldState;
+			});
+		return result.f2 ? result.f3 : null;
+	}
+
+	/**
+	 * Iterator the skip list and perform given function.
+	 *
+	 * @param keyByteBuffer byte buffer storing the key.
+	 * @param keyOffset offset of the key.
+	 * @param keyLen length of the key.
+	 * @param function the function to apply when the skip list contains the given key, which accepts two
+	 *                 parameters: a tuple3 of [previous_node, current_node, next_node] and a boolean indicating
+	 *                 whether the node with same key has been logically removed, and returns a state.
+	 * @return a tuple4 of [previous_node, current_node, key_found, state_by_applying_function]
+	 */
+	private Tuple4<Long, Long, Boolean, S> iterateAndProcess(
```

Comment:
Yes it does, it may break earlier if there's no matching key and key of the iterated node is already bigger than given key, or continue to the end if all existing node keys are smaller than given key. Just check the `if...else if...else` clause at end of the method.